api/v1/writeScreen.js

```diff
@@ -0,0 +1,15 @@
+import { buildThreejsScreen } from '@jsxcad/algorithm-threejs';
+
+export const writeScreen = ({ path }, ...shapes) => {
```

Comment:
As this is part of the API, my first question will be -- what would this do when running in node? :)

Instead of this, I'd suggest writing a geometry to a pathname.

e.g., writePdf({ path: 'gear.stl' }, gear);

This will then work on node.

And then on the browser, using

import { watchFile } from '@jsxcad/sys';
watchFile('gear.stl', renderGeometry); // or something

Now renderGeometry should get called with a 'file' object when the writePdf happens.

The file object should contain 'data' which is the geometry, and a 'translator' which will produce the pdf file content if you call it.

So now, you can convert the 'data' for threejs and display it in the appropriate window for showing a path like 'gear.stl'.
(And call the translate function later on if the user presses a download button or whatever).

So far, this should already exist, but there are a couple of missing bits.

There should be a way to watch for new files.

Then you could say something like watchNewFiles(watcher); and have watcher call watchFile above to catch the data.

Which leaves the missing part of how to map files to windows.

This is something that probably makes sense to add to the api -- and this should be safe because it will have compatible behavior in the browser as well as in node.

I'm not sure what the best approach is, but I guess I'd suggest starting with something like a regular expression matcher.

e.g., ".*" to "main window" or "gear/.*" to "gear window".

Maybe something like

addPreview({ pathPattern: "gear/.*", windowName: "gear" })
addPreview({ pathPattern: "cuts/.*", windowName: "cuts" })

...

writePdf({ path: "cuts/profile.pdf", profile);
writeStl({ path: "gear/preview.stl", gear);

The main thing is that the same script should work both on server and in browser -- and I don't see how that's possible with writeScreen.

Reply:
I agree, that would be very cool. It could also be file type agnostic so you could write a pdf and have it preview nicely in the browser or write an stl file and see a 3D shape. That's definitely getting beyond the things I have experience in so I might not be the right person to do it well. Also since my use case is strictly browser based I'm probably not the best person for the job because I don't really know what a node user would want or not want...but in theory I see what you are saying and I agree that would be the right approach. Obviously we need a solution which works in both node and the browser.